text/0000-std-aware-cargo.md

```diff
@@ -0,0 +1,289 @@
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+> Why is this design the best in the space of possible designs?
+
+This approach borrows from existing behaviors used by Cargo to allow configuration of `core` and `std`, as if they were a regular crate dependency.
+
+This approach also offers an approach that can be developed and applied incrementally, allowing for time to find coner cases not considered by this RFC
+
+> What other designs have been considered and what is the rationale for not choosing them?
+
+To the author of this RFC's knowledge, there are no other open designs, other than the use tools that wrap Cargo entirely, such as [xargo].
+
+[xargo]: https://github.com/japaric/xargo
+
+> What is the impact of not doing this?
+
+By not doing this, Rust will continue to be difficult to use for users and platforms "on the edge", such as new platform developers or embedded and WASM users.
+
+# Prior art
+[prior-art]: #prior-art
+
+* https://github.com/rust-lang/rfcs/pull/1133
+* https://github.com/japaric/xargo
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+## How are dependencies for `core` and `std` specified?
+
+For example in a `no_core` or `no_std` crate, how would we tell Cargo **not** to build the `core` and/or `std` dependencies?
+
+## Should `std` be rebuilt if `core` is rebuilt?
+
+Is it necessary to rebuild `std` using the customized `core`, even if no changes to `std` are necessary?
+
+## Should Cargo obtain or verify the source code for `libcore` or `libstd`?
+
+Right now we depend on `rustup` to obtain the correct source code for these libraries, and we rely on the user not to tamper with the contents. Are these reasonable decisions?
+
+## Should the custom built `libcore` and `libstd` reside locally or globally?
+
+e.g., should the build artifacts be placed in `target/`, only usable by this project, or in `.cargo/`, to be possibly reused by multiple projects, if they happen to have the same settings?
+
+## How do we handle `libcore` and `libstd`'s `Cargo.lock` file?
+
+Right now these are built using the global lock file in `rust-lang/rust`. Should this always be true? How should Cargo handle this gracefully?
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+## Unified `core` and `std`
+
+With the mechanisms specified above, it could be possible to remove the concept of `core` and `std` from the user, leaving only `core`.
+
+By using stable feature flags for `std`, we could say that `std` as a crate with `default-features = false` would essentially be `no_core`, or with `features = ["core"]`, we would be the same as `no_std`.
+
+This abstraction may not map to the actual implementation of `libcore` or `libstd`, but instead be an abstraction layer for the end developer.
```

Comment:
Go into how we could architect such an abstraction layer?

Reply:
Perhaps I was unclear, but the "abstraction layer" is that everything except `std` goes away from the user's perspective, and today's `no_core` and `no_std` are just `std` with no/fewer features set.

I call it an abstraction layer, as under the hood it is unlikely (or undesirable) to try and merge `libcore`, `liballoc`, `libstd`, et. al due to a number of reasons (discussed by the libs team at all-hands).

Let me know how much of that you would like me to add to the RFC or expand upon :)

Reply:
> Let me know how much of that you would like me to add to the RFC or expand upon :)

Mostly, I think it would be good to consider how we might architect the facade that `core` and `alloc` would presumably turn into... possibly with some code snippets as examples.